modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/ReflectionMarshallersProvider.java

```diff
@@ -0,0 +1,177 @@
+
+    /** Constructor. */
+    public ReflectionMarshallersProvider() {
+        this.keysMarshallerCache = new MarshallerCache(CACHE_SIZE);
+        this.valuesMarshallerCache = new MarshallerCache(CACHE_SIZE);
+        this.rowMarshallerCache = new MarshallerCache(CACHE_SIZE);
+        this.projectionMarshallerCache = new MarshallerCache(CACHE_SIZE);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Marshaller getKeysMarshaller(
+            MarshallerSchema schema,
+            Mapper<?> mapper,
+            boolean requireAllFields,
+            boolean allowUnmappedFields) {
+
+        MarshallerCacheKey key = new MarshallerCacheKey(schema.schemaVersion(), mapper, requireAllFields, allowUnmappedFields);
+
+        return keysMarshallerCache.put(key, k -> {
+            return Marshaller.createMarshaller(schema.keys(), key.mapper, key.requireAllFields, key.allowUnmappedFields);
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Marshaller getValuesMarshaller(
+            MarshallerSchema schema,
+            Mapper<?> mapper,
+            boolean requireAllFields,
+            boolean allowUnmappedFields) {
+
+        MarshallerCacheKey key = new MarshallerCacheKey(schema.schemaVersion(), mapper, requireAllFields, allowUnmappedFields);
+
+        return valuesMarshallerCache.put(key, k -> {
+            return Marshaller.createMarshaller(schema.values(), key.mapper, key.requireAllFields, key.allowUnmappedFields);
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Marshaller getRowMarshaller(
+            MarshallerSchema schema,
+            Mapper<?> mapper,
+            boolean requireAllFields,
+            boolean allowUnmappedFields) {
+
+        MarshallerCacheKey key = new MarshallerCacheKey(schema.schemaVersion(), mapper, requireAllFields, allowUnmappedFields);
+
+        return rowMarshallerCache.put(key, k -> {
+            return Marshaller.createMarshaller(schema.row(), key.mapper, key.requireAllFields, key.allowUnmappedFields);
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Marshaller getMarshaller(
+            MarshallerColumn[] columns,
+            Mapper<?> mapper,
+            boolean requireAllFields,
+            boolean allowUnmappedFields) {
+
+        MarshallerCacheKey key = new MarshallerCacheKey(columns, mapper, requireAllFields, allowUnmappedFields);
+
+        return projectionMarshallerCache.put(key, k -> {
+            return Marshaller.createMarshaller(k.columns, k.mapper, k.requireAllFields, k.allowUnmappedFields);
+        });
+    }
+
+    private static class MarshallerCache {
+
+        private final Cache<MarshallerCacheKey, Marshaller> cache;
+
+        MarshallerCache(int maximumSize) {
+            cache = Caffeine.newBuilder()
+                    .maximumSize(maximumSize)
+                    .build();
+        }
+
+        Marshaller put(MarshallerCacheKey key, Function<MarshallerCacheKey, Marshaller> func) {
```

Comment:
```suggestion
        Marshaller getOrAdd(MarshallerCacheKey key, Function<MarshallerCacheKey, Marshaller> func) {
```

`put` is quite confusing here.

Reply:
Fixed.